### PR TITLE
[FXML-4440] Introduce a pass that annotates the type of the argument as an attribute

### PR DESCRIPTION
**Pull request by josel-amd**
Create a pass that annotates all func op inputs with their type as attributes.

### CODE REVIEW

File: mlir/include/mlir/Dialect/Func/Transforms/Passes.h [hunk 1]
```diff
@@ -29,6 +29,9 @@ namespace func {
 #define GEN_PASS_DECL
 #include "mlir/Dialect/Func/Transforms/Passes.h.inc"
 
+/// Creates type attributes for func parameters, that mirror the actual type
```
Comment (mgehre-amd) [inline mlir/include/mlir/Dialect/Func/Transforms/Passes.h:32]: If you remove `let constructor = "mlir::func::createAnnotateInputTypesPass()";` from the tablegen file, then a constructor will be auto-generated and included by 
```
#define GEN_PASS_DECL
#include "mlir/Dialect/Func/Transforms/Passes.h.inc"
```
above.

File: mlir/test/Dialect/Func/annotate-types.mlir [hunk 2]
```diff
@@ -0,0 +1,23 @@
+// RUN: mlir-opt %s --split-input-file --annotate-input-types | \
+// RUN: FileCheck %s
```
Comment (mgehre-amd) [inline mlir/test/Dialect/Func/annotate-types.mlir:2]: nit
```suggestion
// RUN: mlir-opt %s --split-input-file --annotate-input-types | FileCheck %s
```

File: mlir/test/Dialect/Func/annotate-types.mlir [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+
+// -----
```
Comment (mgehre-amd) [inline mlir/test/Dialect/Func/annotate-types.mlir:5]:
```suggestion

```

Comment (mgehre-amd) [inline mlir/test/Dialect/Func/annotate-types.mlir:5]:
```suggestion
```

**Comment by mgehre-amd** [comment]
We will also need to remember the types of the function results. To that end, we might need to find a more generic name of this pass, maybe something like `AnnotateFunctionType`?

File: mlir/lib/Dialect/Func/Transforms/AnnotateInputTypes.cpp [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    for (const auto [argNum, type] : llvm::enumerate(inputs)) {
+      func.setArgAttr(argNum, "func.orig_type", TypeAttr::get(type));
+    }
+  }
+};
+} // namespace
+
+std::unique_ptr<Pass> mlir::func::createAnnotateInputTypesPass() {
+  return std::make_unique<AnnotateInputTypesPass>();
+}
```
Comment (cferry-AMD) [inline mlir/lib/Dialect/Func/Transforms/AnnotateInputTypes.cpp:45]:
```suggestion
}

```

File: mlir/include/mlir/Dialect/Func/Transforms/Passes.td [hunk 5]
```diff
@@ -51,4 +51,17 @@ def DuplicateFunctionEliminationPass : Pass<"duplicate-function-elimination",
   let constructor = "mlir::func::createDuplicateFunctionEliminationPass()";
 }
 
+def AnnotateInputTypes : Pass<"annotate-input-types", "mlir::func::FuncOp"> {
```
Comment (cferry-AMD) [inline mlir/include/mlir/Dialect/Func/Transforms/Passes.td:54]:
```suggestion
def AnnotateInputTypes : Pass<"annotate-input-types", "func::FuncOp"> {
```

**Review by cferry-AMD — COMMENTED** [review]
Good idea! Like @mgehre-amd said, we should also save the return types.

File: mlir/include/mlir/Dialect/Func/Transforms/Passes.h (hunk 1 above)
Comment (josel-amd) [reply mlir/include/mlir/Dialect/Func/Transforms/Passes.h:32]: TIL

File: mlir/include/mlir/Dialect/Func/Transforms/Passes.td [hunk 6]
```diff
@@ -51,4 +51,15 @@ def DuplicateFunctionEliminationPass : Pass<"duplicate-function-elimination",
   let constructor = "mlir::func::createDuplicateFunctionEliminationPass()";
 }
 
+def AnnotateFunctionType: Pass<"annotate-function-type", "func::FuncOp"> {
+  let summary = "Annotate the function type as type attributes";
+  let description = [{
+    Annotates all the inputs and outputs of func.func operators with a type
+    attribute. The type attribute mirrors the actual type of the inputs/outputs.
+
+    This pass can be used to trace back the original input types of func.func
```
Comment (mgehre-amd) [inline mlir/include/mlir/Dialect/Func/Transforms/Passes.td:60]:
```suggestion
    This pass can be used to trace back the original types of func.func
```

**Review by mgehre-amd — APPROVED** [review]

File: mlir/test/Dialect/Func/annotate-types.mlir [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+// CHECK-LABEL: func.func @one_arg_int(%arg0: tensor<ui8> {func.orig_type = tensor<ui8>}) -> (tensor<ui8> {func.orig_type = tensor<ui8>}) {
+func.func @one_arg_int(%arg0 : tensor<ui8>) -> tensor<ui8> {
+  return %arg0 : tensor<ui8>
+}
+
+
+// -----
+// CHECK-LABEL: func.func @n_rank_tensor(%arg0: tensor<3x4x5xf32> {func.orig_type = tensor<3x4x5xf32>}) -> (tensor<3x4x5xf32> {func.orig_type = tensor<3x4x5xf32>}) {
+func.func @n_rank_tensor(%arg0 :tensor<3x4x5xf32>) -> tensor<3x4x5xf32> {
```
Comment (TinaAMD) [inline mlir/test/Dialect/Func/annotate-types.mlir:18]:
```suggestion
func.func @n_rank_tensor(%arg0: tensor<3x4x5xf32>) -> tensor<3x4x5xf32> {
```

File: mlir/test/Dialect/Func/annotate-types.mlir [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+// CHECK-LABEL: func.func @one_arg(%arg0: tensor<f32> {func.orig_type = tensor<f32>}) -> (tensor<f32> {func.orig_type = tensor<f32>}) {
+func.func @one_arg(%arg0 : tensor<f32>) -> tensor<f32> {
+  return %arg0 : tensor<f32>
+}
+
+
+// -----
+// CHECK-LABEL: func.func @one_arg_int(%arg0: tensor<ui8> {func.orig_type = tensor<ui8>}) -> (tensor<ui8> {func.orig_type = tensor<ui8>}) {
+func.func @one_arg_int(%arg0 : tensor<ui8>) -> tensor<ui8> {
```
Comment (TinaAMD) [inline mlir/test/Dialect/Func/annotate-types.mlir:11]:
```suggestion
func.func @one_arg_int(%arg0: tensor<ui8>) -> tensor<ui8> {
```

File: mlir/test/Dialect/Func/annotate-types.mlir [hunk 9]
```diff
@@ -0,0 +1,27 @@
+// RUN: mlir-opt %s --split-input-file --annotate-function-type | FileCheck %s
+
+// CHECK-LABEL: func.func @one_arg(%arg0: tensor<f32> {func.orig_type = tensor<f32>}) -> (tensor<f32> {func.orig_type = tensor<f32>}) {
+func.func @one_arg(%arg0 : tensor<f32>) -> tensor<f32> {
```
Comment (TinaAMD) [inline mlir/test/Dialect/Func/annotate-types.mlir:4]:
```suggestion
func.func @one_arg(%arg0: tensor<f32>) -> tensor<f32> {
```

**Review by TinaAMD — APPROVED** [review]
LGTM, only some nitpicks

File: mlir/test/Dialect/Func/annotate-types.mlir [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+  return %arg0 : tensor<f32>
+}
+
+
+// -----
```
Comment (cferry-AMD) [inline mlir/test/Dialect/Func/annotate-types.mlir:9]:
```suggestion
// -----

```

Comment (cferry-AMD) [reply mlir/test/Dialect/Func/annotate-types.mlir:9]: In the rest of LLVM, it looks like the splits leave one line above and one line below each test case.

File: mlir/lib/Dialect/Func/Transforms/AnnotateFunctionType.cpp [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+#include "mlir/Dialect/Func/Transforms/Passes.h"
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/Pass/Pass.h"
+
+using namespace mlir;
+
+namespace mlir::func {
+#define GEN_PASS_DECL_ANNOTATEFUNCTIONTYPE
```
Comment (cferry-AMD) [inline mlir/lib/Dialect/Func/Transforms/AnnotateFunctionType.cpp:24]: This should go into `mlir/Dialect/Func/Transforms/Passes.h` for other passes to use; only the definition needs to be here.

**Review by cferry-AMD — APPROVED** [review]
Looks good!

**Pull request merged**